File: tools/todo-tree.py
#!/usr/bin/env python3
"""Minimal TODO tree manager for HOOTNER project."""
import sys
import json
from pathlib import Path
from typing import List, Dict

def parse(lines: List[str]) -> List[Dict]:
    """Parse indented lines into tree."""
    tree, stack = [], [tree]
    for line in lines:
        s = line.rstrip()
        if not s: continue
        indent = len(line) - len(line.lstrip())
        task = {'text': s.lstrip(), 'sub': [], 'lvl': indent // 4}
        while len(stack) > task['lvl'] + 1: stack.pop()
        stack[-1].append(task)
        stack.append(task['sub'])
    return tree

def show(tasks: List[Dict], pre: str = "") -> None:
    """Display tree."""
    for i, t in enumerate(tasks):
        conn = "└─ " if i == len(tasks) - 1 else "├─ "
        print(f"{pre}{conn}{t['text']}")
        show(t['sub'], pre + ("    " if i == len(tasks) - 1 else "│   "))

def toggle(tasks: List[Dict], kw: str) -> bool:
    """Toggle [ ] <-> [x]."""
    found = False
    for t in tasks:
        if kw.lower() in t['text'].lower():
            if '[x]' in t['text']: t['text'] = t['text'].replace('[x]', '[ ]')
            elif '[ ]' in t['text']: t['text'] = t['text'].replace('[ ]', '[x]')
            else: t['text'] += ' [x]'
            found = True
        found |= toggle(t['sub'], kw)
    return found

def search(tasks: List[Dict], kw: str, res: List[str]) -> None:
    """Find matching tasks."""
    for t in tasks:
        if kw.lower() in t['text'].lower(): res.append(t['text'])
        search(t['sub'], kw, res)

def to_lines(tasks: List[Dict], lvl: int = 0) -> List[str]:
    """Convert tree to indented lines."""
    lines = []
    for t in tasks:
        lines.append('    ' * lvl + t['text'])
        lines.extend(to_lines(t['sub'], lvl + 1))
    return lines

def main():
    if len(sys.argv) < 3:
        print("Usage: python todo-tree.py <file> <show|done|search> [args]")
        sys.exit(1)
    
    file, cmd = Path(sys.argv[1]), sys.argv[2]
    tree = parse(file.read_text().splitlines()) if file.exists() else []
    
    if cmd == 'show':
        show(tree)
    elif cmd == 'done' and len(sys.argv) > 3:
        if toggle(tree, sys.argv[3]):
            file.write_text('\n'.join(to_lines(tree)) + '\n')
            print("✓ Saved")
        else:
            print("✗ No match")
    elif cmd == 'search' and len(sys.argv) > 3:
        res = []
        search(tree, sys.argv[3], res)
        print('\n'.join(res) if res else "✗ No match")
    else:
        print("Commands: show | done <keyword> | search <keyword>")
        sys.exit(1)

if __name__ == '__main__':
    main()

File: tools/todo-tree.js
#!/usr/bin/env node
/**
 * Minimal TODO tree manager for HOOTNER project
 */
const fs = require('fs');

function parse(lines) { const tree = [], stack = [tree];
  for (const line of lines) { const s = line.trimEnd();
    if (!s) continue;
    const indent = line.length - line.trimStart().length;
    const task = { text: s.trimStart(), sub: [], lvl: Math.floor(indent / 4) };
    while (stack.length > task.lvl + 1) stack.pop();
    stack[stack.length - 1].push(task);
    stack.push(task.sub); }
  return tree; }

function show(tasks, pre = '') { tasks.forEach((t, i) => { const conn = i === tasks.length - 1 ? '└─ ' : '├─ ';
        show(t.sub, pre + (i === tasks.length - 1 ? '    ' : '│   ')); }); }

function toggle(tasks, kw) { let found = false;
  for (const t of tasks) { if (t.text.toLowerCase().includes(kw.toLowerCase())) { if (t.text.includes('[x]')) t.text = t.text.replace('[x]', '[ ]');
      else if (t.text.includes('[ ]')) t.text = t.text.replace('[ ]', '[x]');
      else t.text += ' [x]';
      found = true; }
    found |= toggle(t.sub, kw); }
  return found; }

function search(tasks, kw, res = []) { for (const t of tasks) { if (t.text.toLowerCase().includes(kw.toLowerCase())) res.push(t.text);
    search(t.sub, kw, res); }
  return res; }

function toLines(tasks, lvl = 0) { const lines = [];
  for (const t of tasks) { lines.push('    '.repeat(lvl) + t.text);
    lines.push(...toLines(t.sub, lvl + 1)); }
  return lines; }

const [,, file, cmd, ...args] = process.argv;
if (!file || !cmd) { process.exit(1); }

const tree = fs.existsSync(file) ? parse(fs.readFileSync(file, 'utf8').split('\n')) : [];

if (cmd === 'show') { show(tree); } else if (cmd === 'done' && args[0]) { if (toggle(tree, args[0])) { fs.writeFileSync(file, toLines(tree).join('\n') + '\n'); } else { } } else if (cmd === 'search' && args[0]) { const res = search(tree, args[0]);
   : '✗ No match'); } else { process.exit(1); }
